### PR TITLE
Replace `ThreadLocal`s with `ObjectPool`s

### PR DESCRIPTION
**Pull request by overheadhunter**
We used to use `ThreadLocal<Cipher>` instances, so we can reuse Cipher instances, as creating new Ciphers is fairly expensive due to ServiceLoaders involved (especially on older JREs).

Normally, if a thread is no longer alive, the ThreadLocals are removed. However, if the thread is re-used, the ThreadLocal may leak.

This PR replaces ThreadLocals with [ObjectPool](https://github.com/cryptomator/cryptolib/blob/a0954d48548da1701aa3d37dde1bb200271636ce/src/main/java/org/cryptomator/cryptolib/common/ObjectPool.java)s which allow rapid re-use of Ciphers (even across threads) while still allowing them to be garbage collected.

This comes with a small performance penalty, though:

```
Benchmark                    Mode  Cnt  Score   Error  Units
Uncached createCipher        avgt    2  6,045          us/op
Uncached createMac           avgt    2  1,058          us/op
Uncached createMd            avgt    2  0,159          us/op
ThreadLocal createCipher     avgt    2  1,455          us/op
ThreadLocal createMac        avgt    2  0,186          us/op
ThreadLocal createMd         avgt    2  0,085          us/op
Pooled recycleCipher         avgt    2  1,753          us/op
Pooled recycleMac            avgt    2  0,401          us/op
Pooled recycleMd             avgt    2  0,244          us/op
```

### CODE REVIEW

**Comment by overheadhunter** [comment]
@SailReal This change deprecates the old API (while retaining it for now) and adds a new alternative. I can not say if this affects Android at all, but I'd rather like to hear your opinion whether it is feasible to switch to the new API.

**Comment by SailReal** [comment]
Thanks for the information!
Will check it tomorrow and also test it on some some android version to make sure it is compatible.

**Review by SailReal — APPROVED** [review]
LGTM :+1: and tested using our min/max version (Android 7.0/12.0) with vault format 7/8 
* create/add a vault
* create/open a file/folder

File: src/main/java/org/cryptomator/cryptolib/v1/FileContentCryptorImpl.java [hunk 1]
```diff
@@ -141,9 +143,10 @@ void decryptChunk(ByteBuffer ciphertextChunk, ByteBuffer cleartextChunk, Destroy
...
-			final Cipher cipher = CipherSupplier.AES_CTR.forDecryption(fk, new IvParameterSpec(nonce));
-			assert cleartextChunk.remaining() >= cipher.getOutputSize(payloadBuf.remaining());
-			cipher.doFinal(payloadBuf, cleartextChunk);
+			try (ObjectPool.Lease<Cipher> cipher = CipherSupplier.AES_CTR.decryptionCipher(fk, new IvParameterSpec(nonce))) {
+				assert cleartextChunk.remaining() >= cipher.get().getOutputSize(payloadBuf.remaining());
+				cipher.get().doFinal(payloadBuf, cleartextChunk);
+			}
 		} catch (ShortBufferException e) {
 			throw new IllegalStateException("Buffer allocated for reported output size apparently not big enough.", e);
 		} catch (IllegalBlockSizeException | BadPaddingException e) {
```
[the remark this answers is not in the document]
Comment (overheadhunter) [reply src/main/java/org/cryptomator/cryptolib/v1/FileContentCryptorImpl.java:152]: @sonatype-lift ignore false positive (try-with-resource)

File: src/main/java/org/cryptomator/cryptolib/v1/FileHeaderCryptorImpl.java [hunk 2]
```diff
@@ -100,12 +103,12 @@ public FileHeader decryptHeader(ByteBuffer ciphertextHeaderBuf) throws Authentic
 		buf.get(expectedMac);
 
 		// check mac:
-		try (DestroyableSecretKey mk = masterkey.getMacKey()) {
+		try (DestroyableSecretKey mk = masterkey.getMacKey();
```
[the remark this answers is not in the document]
Comment (overheadhunter) [reply src/main/java/org/cryptomator/cryptolib/v1/FileHeaderCryptorImpl.java:106]: @sonatype-lift ignore false positive (try-with-resource)

File: src/main/java/org/cryptomator/cryptolib/v2/FileContentCryptorImpl.java [hunk 3]
```diff
@@ -103,13 +104,14 @@ void encryptChunk(ByteBuffer cleartextChunk, ByteBuffer ciphertextChunk, long ch
...
-			cipher.doFinal(cleartextChunk, ciphertextChunk);
+			try (ObjectPool.Lease<Cipher> cipher = CipherSupplier.AES_GCM.encryptionCipher(fk, new GCMParameterSpec(GCM_TAG_SIZE * Byte.SIZE, nonce))) {
+				final byte[] chunkNumberBigEndian = longToBigEndianByteArray(chunkNumber);
+				cipher.get().updateAAD(chunkNumberBigEndian);
+				cipher.get().updateAAD(headerNonce);
+				ciphertextChunk.put(nonce);
+				assert ciphertextChunk.remaining() >= cipher.get().getOutputSize(cleartextChunk.remaining());
+				cipher.get().doFinal(cleartextChunk, ciphertextChunk);
+			}
 		} catch (ShortBufferException e) {
```
[the remark this answers is not in the document]
Comment (overheadhunter) [reply src/main/java/org/cryptomator/cryptolib/v2/FileContentCryptorImpl.java:115]: @sonatype-lift ignore false positive (try-with-resource)

File: src/main/java/org/cryptomator/cryptolib/v2/FileContentCryptorImpl.java [hunk 4]
```diff
@@ -132,12 +134,13 @@ void decryptChunk(ByteBuffer ciphertextChunk, ByteBuffer cleartextChunk, long ch
...
+			try (ObjectPool.Lease<Cipher> cipher = CipherSupplier.AES_GCM.decryptionCipher(fk, new GCMParameterSpec(GCM_TAG_SIZE * Byte.SIZE, nonce))) {
+				final byte[] chunkNumberBigEndian = longToBigEndianByteArray(chunkNumber);
+				cipher.get().updateAAD(chunkNumberBigEndian);
+				cipher.get().updateAAD(headerNonce);
+				assert cleartextChunk.remaining() >= cipher.get().getOutputSize(payloadBuf.remaining());
+				cipher.get().doFinal(payloadBuf, cleartextChunk);
+			}
 		} catch (AEADBadTagException e) {
 			throw new AuthenticationFailedException("Content tag mismatch.", e);
 		} catch (ShortBufferException e) {
```
[the remark this answers is not in the document]
Comment (overheadhunter) [reply src/main/java/org/cryptomator/cryptolib/v2/FileContentCryptorImpl.java:146]: @sonatype-lift ignore false positive (try-with-resource)

File: src/main/java/org/cryptomator/cryptolib/v2/FileHeaderCryptorImpl.java [hunk 5]
```diff
@@ -61,10 +62,10 @@ public ByteBuffer encryptHeader(FileHeader header) {
...
-			int encrypted = cipher.doFinal(payloadCleartextBuf, result);
-			assert encrypted == FileHeaderImpl.PAYLOAD_LEN + FileHeaderImpl.TAG_LEN;
-
+			try (ObjectPool.Lease<Cipher> cipher = CipherSupplier.AES_GCM.encryptionCipher(ek, new GCMParameterSpec(GCM_TAG_SIZE * Byte.SIZE, headerImpl.getNonce()))) {
+				int encrypted = cipher.get().doFinal(payloadCleartextBuf, result);
+				assert encrypted == FileHeaderImpl.PAYLOAD_LEN + FileHeaderImpl.TAG_LEN;
+			}
 			result.flip();
 			return result;
 		} catch (ShortBufferException e) {
```
[the remark this answers is not in the document]
Comment (overheadhunter) [reply src/main/java/org/cryptomator/cryptolib/v2/FileHeaderCryptorImpl.java:71]: @sonatype-lift ignore false positive (try-with-resource)

File: src/main/java/org/cryptomator/cryptolib/v1/FileHeaderCryptorImpl.java [hunk 6]
```diff
@@ -100,12 +103,12 @@ public FileHeader decryptHeader(ByteBuffer ciphertextHeaderBuf) throws Authentic
```
[the remark this answers is not in the document]
Comment (overheadhunter) [reply src/main/java/org/cryptomator/cryptolib/v1/FileHeaderCryptorImpl.java]: @sonatype-lift ignore outdated

**Pull request merged**